### PR TITLE
Optimize overrideTags for objects

### PR DESCRIPTION
**Pull request by wesmoncrief**
A couple of optimizations here. These work for when you have tags as objects, not arrays.

1. Often, you only have global tags, and no child tags. This memoizes that case.
2. `overrideTagsUnoptimized` is slow because it does a lot of string manipulation - it serializes then unserializes tag/value pairs. `overrideTags2` can avoid this by only taking in objects. So this optimization is only able to be used when the caller passes in objects (instead of arrays).
3. if there's no intersection between child and global tags, use a memoized global tags string
4. String concat is a bit faster than making an array, then string joining

benchmark results:
```
overrideTagsUnoptimized performance benchmark: 23.790959 ms
overrideTagsUnoptimized for non-pre-made arrays performance benchmark: 19.103083 ms
overrideTags2 performance benchmark: 12.292041 ms
```

### CODE REVIEW

File: lib/statsd.js [hunk 1]
```diff
@@ -242,22 +248,41 @@ Client.prototype.sendStat = function (stat, value, type, sampleRate, tags, callb
...
+  const isTagsArray = Array.isArray(tags);
+  if (isTagsArray) {
+    if (tags.length === 0) {
+      return globalTagsMemo;
+    }
+  }
+  else if (Object.keys(tags).length === 0) {
+    return globalTagsMemo;
+  }
+  if (isTagsArray || Array.isArray(globalTags)) {
```
Comment (donhcd) [inline lib/statsd.js]: the lack of typing here (not really your fault) makes it really hard to follow what the assumptions here are... to is it possible for `tags` to be an array but for `globalTags` to be an object? or the other way around?

does `overrideTagsToString` work if you pass in objects for `globalTags` or `tags`?

Comment (wesmoncrief) [reply lib/statsd.js]: tags could be array or object. globalTags could be array or object. Welcome to the wild west.

`overrideTagsToString` does work for any combination of objects/arrays. 

`overrideTags2` only works if globalTags is an object, and also tags is an object

Comment (wesmoncrief) [reply lib/statsd.js]: btw, the unit tests seem pretty good, but i was thinking i'd also test this by testing on a manual image deployment of xover in a small staging region

File: lib/helpers.js [hunk 2]
```diff
@@ -61,6 +63,58 @@ function overrideTags (parent, child, telegraf) {
...
+    addToResult(serializedTagWithValue);
+  }
+  if (usedGlobalKeys.size === 0) {
+    addToResult(globalTagsFullMemo, true);
+  }
+  else {
+    for (const g of Object.keys(globalTags)) {
+      if (!usedGlobalKeys.has(g)) {
+        const serializedTagWithValue = `${g}:${sanitizeTags(globalTags[g], telegraf)}`;
+        result = serializedTagWithValue + separator + result;
```
Comment (donhcd) [inline lib/helpers.js]: should we use `addToResult(serializedTagWithValue, true)` here?

**Review by donhcd — APPROVED** [review]
LGTM

File: lib/helpers.js (hunk 2 above)
Comment (wesmoncrief) [reply lib/helpers.js]: ya
